### PR TITLE
[API stabilization] JSON configuration becomes stricter

### PR DESCRIPTION
**Pull request by qwwdfsad**
* Allow only whitespaces in prettyPrintIndent
    * Fail on non-finite fp on decoding if allowSpecialFloatingPointValues is disabled
    * Unify exception messages

### CODE REVIEW

File: runtime/commonMain/src/kotlinx/serialization/json/JsonExceptions.kt [hunk 1]
```diff
@@ -28,22 +28,30 @@ internal class JsonEncodingException(message: String) : JsonException(message)
...
-internal fun InvalidFloatingPoint(value: Number, key: String, type: String, output: String) = JsonEncodingException(
-    "'$value' with key '$key' is not a valid $type as per JSON specification. " +
-            "You can enable 'serializeSpecialFloatingPointValues' property to serialize such values.\n" +
+internal fun InvalidFloatingPointDecoded(value: Number, key: String, output: String) =
+    JsonDecodingException(-1, invalidFp(value, key, output))
+
+internal fun InvalidFloatingPoint(value: Number, key: String, output: String) =
+    JsonEncodingException(invalidFp(value, key, output))
+
+private fun invalidFp(value: Number, key: String, output: String): String {
```
Comment (sandwwraith) [inline runtime/commonMain/src/kotlinx/serialization/json/JsonExceptions.kt:44]: This confuses me a lot. I've counted three different ways to throw an exception with 3 different messages: `InvalidFloatingPoint`, `InvalidFloatingPointDecoded`, and `reader.fail` (and strange `invalidFp`). I don't get the logic why you choose one or another. Maybe it's possible to collect them all in one place and name `InvalidFloatingPointEncoded` and `InvalidFloatingPointDecoded`? (with all necessary overloads)

File: runtime/commonTest/src/kotlinx/serialization/json/JsonConfigurationTest.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+package kotlinx.serialization.json
+
+import kotlin.test.*
+
+class JsonConfigurationTest {
+
+    @Test
+    fun testPrettyPrint() {
+        json(true, "")
```
Comment (sandwwraith) [inline runtime/commonTest/src/kotlinx/serialization/json/JsonConfigurationTest.kt:13]: Can you add tests on the correctness of floating point-related exceptions?

File: runtime/commonMain/src/kotlinx/serialization/json/JsonExceptions.kt (hunk 1 above)
Comment (qwwdfsad) [reply runtime/commonMain/src/kotlinx/serialization/json/JsonExceptions.kt:44]: Sure, done

File: runtime/commonTest/src/kotlinx/serialization/json/JsonConfigurationTest.kt (hunk 2 above)
Comment (qwwdfsad) [reply runtime/commonTest/src/kotlinx/serialization/json/JsonConfigurationTest.kt:13]: We already have `SpecialFloatingPointValuesTest`

File: runtime/commonMain/src/kotlinx/serialization/json/internal/JsonExceptions.kt [hunk 3]
```diff
@@ -28,22 +29,37 @@ internal class JsonEncodingException(message: String) : JsonException(message)
...
+    JsonEncodingException(unexpectedFpErrorMessage(value, key, output))
+
+internal fun InvalidFloatingPointDecoded(value: Number, key: String, output: String) =
+    JsonDecodingException(-1, unexpectedFpErrorMessage(value, key, output))
+
+// Extension on JSON reader and fail immediately
+internal fun JsonReader.throwInvalidFloatingPointDecoded(result: Number): Nothing {
+    fail("Unexpected special floating-point value $result. By default, " +
+            "non-finite floating point values are prohibited because they do not conform JSON specification. " +
+            "It is possible to serialize them using 'JsonBuilder.allowSpecialFloatingPointValues = true'")
```
Comment (sandwwraith) [inline runtime/commonMain/src/kotlinx/serialization/json/internal/JsonExceptions.kt:49]:
```suggestion
            "It is possible to deserialize them using 'JsonBuilder.allowSpecialFloatingPointValues = true'")
```

File: runtime/commonMain/src/kotlinx/serialization/json/internal/JsonExceptions.kt [hunk 4]
```diff
@@ -28,22 +29,37 @@ internal class JsonEncodingException(message: String) : JsonException(message)
...
+}
+
+private fun unexpectedFpErrorMessage(value: Number, key: String, output: String): String {
+    return "Unexpected special floating-point value $value with key $key. By default, " +
+            "non-finite floating point values are prohibited because they do not conform JSON specification. " +
+            "It is possible to serialize them using 'JsonBuilder.allowSpecialFloatingPointValues = true'\n" +
```
Comment (sandwwraith) [inline runtime/commonMain/src/kotlinx/serialization/json/internal/JsonExceptions.kt:55]:
```suggestion
            "It is possible to deserialize them using 'JsonBuilder.allowSpecialFloatingPointValues = true'\n" +
```

**Pull request merged**